crates/db_schema/src/utils.rs

```diff
@@ -580,102 +572,6 @@ impl<T: Expression + AsExpression<sql_types::Record<T::SqlType>>> AsRecord for T
-///
-/// This example modifies a query to make columns in `community_actions` available:
-///
-/// ```
-/// community::table
-///   .left_join(actions(
-///     community_actions::table,
-///     my_person_id,
-///     community::id,
-///   ))
-/// ```
-pub fn actions<T, P, C, K0, K1>(
-  actions_table: T,
-  person_id: Option<P>,
-  target_id: C,
-) -> OnTupleEq<T, dsl::Nullable<K0>, K1, BindIfSome<dsl::AsExprOf<P, sql_types::Integer>>, C>
-where
-  T: Table<PrimaryKey = (K0, K1)> + Copy,
-  K0: Expression,
-  P: AsExpression<sql_types::Integer>,
-  (dsl::Nullable<K0>, K1): AsRecord,
-  (BindIfSome<dsl::AsExprOf<P, sql_types::Integer>>, C):
-    AsExpression<<AsRecordOutput<(dsl::Nullable<K0>, K1)> as Expression>::SqlType>,
-{
-  let (k0, k1) = actions_table.primary_key();
-  actions_table.on((k0.nullable(), k1).into_sql().eq((
-    BindIfSome(person_id.map(diesel::IntoSql::into_sql)),
-    target_id,
-  )))
-}
-
-/// Like `actions` but `actions_table` is an alias and person id is not nullable
-#[allow(clippy::type_complexity)]
-pub fn actions_alias<T, P, C, K0, K1>(
-  actions_table: Alias<T>,
-  person_id: P,
-  target_id: C,
-) -> OnTupleEq<Alias<T>, AliasedField<T, K0>, AliasedField<T, K1>, P, C>
-where
-  Alias<T>: QuerySource + Copy,
-  T: AliasSource<Target: Table<PrimaryKey = (K0, K1)>> + Default,
-  K0: Column<Table = T::Target>,
-  K1: Column<Table = T::Target>,
-  (AliasedField<T, K0>, AliasedField<T, K1>): AsRecord,
-  (P, C): AsExpression<
-    <AsRecordOutput<(AliasedField<T, K0>, AliasedField<T, K1>)> as Expression>::SqlType,
-  >,
-{
-  let (k0, k1) = T::default().target().primary_key();
-  actions_table.on(
-    (actions_table.field(k0), actions_table.field(k1))
-      .into_sql()
-      .eq((person_id, target_id)),
-  )
-}
-
-/// `action_query(table_name::action_name)` is the same as
-/// `table_name::table.filter(table_name::action_name.is_not_null())`.
-pub fn action_query<C>(column: C) -> dsl::Filter<C::Table, dsl::IsNotNull<C>>
-where
-  C: Column<Table: Default + FilterDsl<dsl::IsNotNull<C>>, SqlType: SingleValue>,
-{
-  action_query_with_fn(column, |t| t)
-}
-
-/// `find_action(table_name::action_name, key)` is the same as
-/// `table_name::table.find(key).filter(table_name::action_name.is_not_null())`.
-pub fn find_action<C, K>(
-  column: C,
-  key: K,
-) -> dsl::Filter<dsl::Find<C::Table, K>, dsl::IsNotNull<C>>
-where
-  C:
-    Column<Table: Default + FindDsl<K, Output: FilterDsl<dsl::IsNotNull<C>>>, SqlType: SingleValue>,
-{
-  action_query_with_fn(column, |t| t.find(key))
-}
-
-/// `action_query_with_fn(table_name::action_name, f)` is the same as
-/// `f(table_name::table).filter(table_name::action_name.is_not_null())`.
-fn action_query_with_fn<C, Q>(
-  column: C,
-  f: impl FnOnce(C::Table) -> Q,
-) -> dsl::Filter<Q, dsl::IsNotNull<C>>
-where
-  C: Column<Table: Default, SqlType: SingleValue>,
-  Q: FilterDsl<dsl::IsNotNull<C>>,
-{
-  f(C::Table::default()).filter(column.is_not_null())
-}
-
```

Comment:
We should at least keep find_action and action_query.

The actions and actions_query functions might work with the auto_type macro if a matching type alias is made for their return type. The docs for auto_type say that it does this:

> Assume that functions, unless otherwise annotated, have a type alias for their return type available at the same path as the function itself (including case). (e.g. for the `dsl::not(x)` call, it expects that there is a `dsl::not<X>` type alias available)

You might also need put each `Option<PersonId>` in a type-annotated variable.

Reply:
They don't' work with auto_type unfortunately, no generic function does. You can give it a try, but I'd rather they be strictly typed anyway.